### PR TITLE
feat(dynamic_log_level): support timeout for dynamic log level

### PR DESCRIPTION
**Pull request by ADD-SP**
Sister PR: https://github.com/Kong/lua-kong-nginx-module/pull/60

_Fix [KAG-457](https://konghq.atlassian.net/browse/KAG-457), [KAG-5](https://konghq.atlassian.net/browse/KAG-5)_

[KAG-457]: https://konghq.atlassian.net/browse/KAG-457?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[KAG-5]: https://konghq.atlassian.net/browse/KAG-5?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: build/openresty/patches/nginx-1.21.4_06-dynamic_log_level.patch [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+--- a/bundle/nginx-1.21.4/src/core/ngx_log.c
++++ b/bundle/nginx-1.21.4/src/core/ngx_log.c
+@@ -8,6 +8,10 @@
+ #include <ngx_config.h>
+ #include <ngx_core.h>
+ 
++#if (NGX_HTTP_LUA_KONG)
++ngx_uint_t
++ngx_http_lua_kong_get_dynamic_log_level(ngx_uint_t current_log_level);
++#endif
```
Comment (chronolaw) [inline build/openresty/patches/nginx-1.21.4_06-dynamic_log_level.patch:12]: We have included `ngx_log.h` in `ngx_core.h`, so we should remove these code.

File: build/openresty/patches/nginx-1.21.4_06-dynamic_log_level.patch [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+ 
+ #define ngx_log_error(level, log, args...)                                    \
+-    if ((log)->log_level >= level) ngx_log_error_core(level, log, args)
++    if (ngx_http_lua_kong_get_dynamic_log_level((log)->log_level) >= level) ngx_log_error_core(level, log, args)
+ 
+ void ngx_log_error_core(ngx_uint_t level, ngx_log_t *log, ngx_err_t err,
+     const char *fmt, ...);
+ 
+ #define ngx_log_debug(level, log, args...)                                    \
+-    if ((log)->log_level & level)                                             \
```
Comment (chronolaw) [inline build/openresty/patches/nginx-1.21.4_06-dynamic_log_level.patch:85]: forgot something here?

File: kong/api/routes/debug.lua [hunk 3]
```diff
@@ -19,27 +22,31 @@ local function handle_put_log_level(self, broadcast)
...
+  local cur_log_level = get_log_level(LOG_LEVELS[kong.configuration.log_level])
 
-  if sys_filter_level == log_level then
+  if cur_log_level == log_level then
     local message = "log level is already " .. self.params.log_level
     return kong.response.exit(200, { message = message })
   end
 
-  local ok, err = pcall(set_log_level, log_level)
+  local ok, err = pcall(set_log_level, log_level, 30)
```
Comment (chronolaw) [inline kong/api/routes/debug.lua:38]: magic number `30`

File: build/openresty/patches/nginx-1.21.4_06-dynamic_log_level.patch (hunk 1 above)
Comment (ADD-SP) [reply build/openresty/patches/nginx-1.21.4_06-dynamic_log_level.patch:12]: This is a `.c` file, not a header file.

Comment (chronolaw) [reply build/openresty/patches/nginx-1.21.4_06-dynamic_log_level.patch:12]: In this patch we have declared `ngx_http_lua_kong_get_dynamic_log_level` in `ngx_log.h`, so we should not repeat it again.

File: build/openresty/patches/nginx-1.21.4_06-dynamic_log_level.patch [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+index eb7a989..0862d4d 100644
+--- a/bundle/nginx-1.21.4/src/core/ngx_log.c
++++ b/bundle/nginx-1.21.4/src/core/ngx_log.c
+@@ -171,8 +171,12 @@ ngx_log_error_core(ngx_uint_t level, ngx_log_t *log, ngx_err_t err,
+     debug_connection = (log->log_level & NGX_LOG_DEBUG_CONNECTION) != 0;
+
+     while (log) {
+-
++#if (NGX_HTTP_LUA_KONG)
++        if (ngx_http_lua_kong_get_dynamic_log_level(log->log_level) < level &&
```
Comment (chronolaw) [inline build/openresty/patches/nginx-1.21.4_06-dynamic_log_level.patch:11]: As we talked offline, this is a patch for the core module, but here we invoke a *http* function, it makes the code a little weird.

File: kong/api/routes/debug.lua [hunk 5]
```diff
@@ -19,27 +23,31 @@ local function handle_put_log_level(self, broadcast)
...
+  local ok, err = pcall(set_log_level, log_level, timeout)
 
   if not ok then
     local message = "failed setting log level: " .. err
     return kong.response.exit(500, { message = message })
   end
 
   -- broadcast to all workers in a node
-  ok, err = kong.worker_events.post("debug", "log_level", log_level)
+  ok, err = kong.worker_events.post("debug", "log_level", {
```
Comment (chronolaw) [inline kong/api/routes/debug.lua:47]: could we create a table for the event data,
```
local data = { log_level = log_level, timeout = timeout,}
```

File: kong/api/routes/debug.lua [hunk 6]
```diff
@@ -57,28 +67,35 @@ local function handle_put_log_level(self, broadcast)
...
   ["/debug/node/log-level"] = {
     GET = function(self)
-      local sys_filter_level = get_sys_filter_level()
-      local cur_level = LOG_LEVELS[sys_filter_level]
+      local cur_level = get_log_level(LOG_LEVELS[kong.configuration.log_level])
 
-      if type(cur_level) ~= "string" then
-        local message = "unknown log level: " .. tostring(sys_filter_level)
+      if type(LOG_LEVELS[cur_level]) ~= "string" then
+        local message = "unknown log level: " .. tostring(LOG_LEVELS[cur_level])
```
Comment (chronolaw) [inline kong/api/routes/debug.lua:94]:
```suggestion
        local message = "unknown log level: " .. tostring(cur_level)
```

Comment (ADD-SP) [reply kong/api/routes/debug.lua]: done

File: kong/runloop/handler.lua [hunk 7]
```diff
@@ -106,6 +107,7 @@ local set_upstream_ssl_trusted_store = ktls.set_upstream_ssl_trusted_store
 if is_http_module then
   set_authority = require("resty.kong.grpc").set_authority
   set_log_level = require("resty.kong.log").set_log_level
+  get_log_level = require("resty.kong.log").get_log_level
```
Comment (chronolaw) [inline kong/runloop/handler.lua:110]: Perhaps after this PR is merged, we could try to enable them in stream subsystem.

Comment (ADD-SP) [reply kong/runloop/handler.lua:110]: sure

**Review by chronolaw — APPROVED** [review]
When lua-kong-module releases the new version, this PR is ready to merge.

**Comment by chronolaw** [comment]
Should we add a change log entry to describe this new feature?

**Review by attenuation — APPROVED** [review]

File: kong/api/routes/debug.lua [hunk 8]
```diff
@@ -1,16 +1,20 @@
...
+local kong                        = kong
+local pcall                       = pcall
+local type                        = type
+local tostring                    = tostring
+
+local get_log_level               = require("resty.kong.log").get_log_level
+
+local NODE_LEVEL_BROADCAST        = false
+local CLUSTER_LEVEL_BROADCAST     = true
+local DEFAULT_LOG_LEVEL_TIMEOUT   = 60 * 10 -- 10 minutes
```
Comment (fffonion) [inline kong/api/routes/debug.lua:16]: suggest adding unit into the variable name, like `DEFAULT_LOG_LEVEL_TIMEOUT_SECONDS`

**Review by fffonion — APPROVED** [review]

File: autodoc/admin-api/data/admin-api.lua [hunk 9]
```diff
@@ -737,6 +737,9 @@ return {
           description = [[
             Change the log level of a node.
 
+            You can set the timeout in seconds with querystring parameter `timeout`,
```
[the remark this answers is not in the document]
Comment (ADD-SP) [reply autodoc/admin-api/data/admin-api.lua]: done

File: kong/api/routes/debug.lua [hunk 10]
```diff
@@ -57,28 +67,35 @@ local function handle_put_log_level(self, broadcast)
...
   -- store in shm so that newly spawned workers can update their log levels
-  ok, err = ngx.shared.kong:set("kong:log_level", log_level)
+  ok, err = ngx.shared.kong:set("kong:log_level", log_level, timeout)
 
   if not ok then
     local message = "failed storing log level in shm: " .. err
     return kong.response.exit(500, { message = message })
   end
 
+  ok, err = ngx.shared.kong:set("kong:log_level_timeout_at", ngx.time() + timeout, timeout)
```
[the remark this answers is not in the document]
Comment (ADD-SP) [reply kong/api/routes/debug.lua]: done

File: kong/runloop/handler.lua [hunk 11]
```diff
@@ -887,10 +889,11 @@ return {
       if is_http_module then
         -- if worker has outdated log level (e.g. newly spawned), updated it
         timer_at(0, function()
-          local cur_log_level = get_sys_filter_level()
+          local cur_log_level = get_log_level(LOG_LEVELS[kong.configuration.log_level])
           local shm_log_level = ngx.shared.kong:get("kong:log_level")
-          if cur_log_level and shm_log_level and cur_log_level ~= shm_log_level then
-            local ok, err = pcall(set_log_level, shm_log_level)
+          local timeout = (tonumber(ngx.shared.kong:get("kong:log_level_timeout_at")) or 0) - ngx.time()
```
[the remark this answers is not in the document]
Comment (ADD-SP) [reply kong/runloop/handler.lua]: done

File: spec/02-integration/04-admin_api/22-debug_spec.lua [hunk 12]
```diff
@@ -493,9 +523,46 @@ describe("Admin API - Kong debug route with strategy #" .. strategy, function()
         json = cjson.decode(body)
 
         message = "log level: crit"
-        return json.message == message
+        assert.same(message, json.message)
       end, 30)
     end)
+
+    it("change log_level with timeout", function()
```
[the remark this answers is not in the document]
Comment (ADD-SP) [reply spec/02-integration/04-admin_api/22-debug_spec.lua:542]: done

**Comment by outsinre** [comment]
Will cherry-pick by @ADD-SP .

File: autodoc/admin-api/data/admin-api.lua [hunk 13]
```diff
@@ -737,6 +737,14 @@ return {
           description = [[
             Change the log level of a node.
 
+            #### Request Querystring Parameters
+
+            Attributes | Description
+            ---:| ---
+            `timeout`<br>*optional* | The timeout for dynamic log_level, after that, the log level will be reset to the
+            default `log_level` setting from Nginx configuration immediately. If it is set to `0`, the dynamic log_level
+            will exprie immediately. Defaults to `600`.
```
Comment (chronolaw) [inline autodoc/admin-api/data/admin-api.lua:746]:
```suggestion
            will expire immediately. Defaults to `600`.
```

File: autodoc/admin-api/data/admin-api.lua [hunk 14]
```diff
@@ -775,6 +783,14 @@ return {
           description = [[
             Change the log level of all nodes in a cluster.
 
+            #### Request Querystring Parameters
+
+            Attributes | Description
+            ---:| ---
+            `timeout`<br>*optional* | The timeout for dynamic log_level, after that, the log level will be reset to the
+            default `log_level` setting from Nginx configuration immediately. If it is set to `0`, the dynamic log_level
+            will exprie immediately. Defaults to `600`.
```
Comment (chronolaw) [inline autodoc/admin-api/data/admin-api.lua:792]:
```suggestion
            will expire immediately. Defaults to `600`.
```

File: autodoc/admin-api/data/admin-api.lua [hunk 15]
```diff
@@ -821,6 +837,14 @@ return {
             Change the log level of all Control Plane nodes deployed in Hybrid
             (CP/DP) cluster.
 
+            #### Request Querystring Parameters
+
+            Attributes | Description
+            ---:| ---
+            `timeout`<br>*optional* | The timeout for dynamic log_level, after that, the log level will be reset to the
+            default `log_level` setting from Nginx configuration immediately. If it is set to `0`, the dynamic log_level
+            will exprie immediately. Defaults to `600`.
```
Comment (chronolaw) [inline autodoc/admin-api/data/admin-api.lua:846]:
```suggestion
            will expire immediately. Defaults to `600`.
```

File: kong/api/routes/debug.lua [hunk 16]
```diff
@@ -1,16 +1,22 @@
-local get_sys_filter_level = require("ngx.errlog").get_sys_filter_level
-local set_log_level = require("resty.kong.log").set_log_level
+local set_log_level                = require("resty.kong.log").set_log_level
+local cjson                        = require("cjson.safe")
 
-local LOG_LEVELS = require("kong.constants").LOG_LEVELS
+local LOG_LEVELS                   = require("kong.constants").LOG_LEVELS
```
Comment (chronolaw) [inline kong/api/routes/debug.lua:4]: Should we add a single line?
```
local constants = require("kong.constants")
```

**Pull request merged**